src/storage/buffer_manager/buffer_manager.cpp

```diff
@@ -70,47 +65,85 @@ BufferManager::~BufferManager() = default;
 // both get access to the same piece of memory.
 uint8_t* BufferManager::pin(
     BMFileHandle& fileHandle, common::page_idx_t pageIdx, PageReadPolicy pageReadPolicy) {
-    fileHandle.acquirePageLock(pageIdx, LockMode::SPIN);
     auto pageState = fileHandle.getPageState(pageIdx);
-    if (pageState->isInFrame()) {
-        pageState->incrementPinCount();
-    } else {
-        if (!claimAFrame(fileHandle, pageIdx, pageReadPolicy)) {
-            pageState->releaseLock();
-            throw BufferManagerException("Failed to claim a frame.");
+    while (true) {
+        auto currStateAndVersion = pageState->getStateAndVersion();
+        switch (PageState::getState(currStateAndVersion)) {
+        case PageState::EVICTED: {
+            if (pageState->tryLock(currStateAndVersion)) {
+                if (!claimAFrame(fileHandle, pageIdx, pageReadPolicy)) {
+                    pageState->unlock();
+                    throw BufferManagerException("Failed to claim a frame.");
+                }
+                return getFrame(fileHandle, pageIdx);
+            }
+        } break;
+        case PageState::UNLOCKED:
+        case PageState::MARKED: {
+            if (pageState->tryLock(currStateAndVersion)) {
+                return getFrame(fileHandle, pageIdx);
+            }
+        } break;
+        case PageState::LOCKED: {
+            continue;
+        }
+        }
+    }
+}
+
+void BufferManager::optimisticRead(BMFileHandle& fileHandle, common::page_idx_t pageIdx,
+    const std::function<void(uint8_t*)>& func) {
+    auto pageState = fileHandle.getPageState(pageIdx);
+    while (true) {
```

Comment:
This is a piece of code that makes me appreciate papers (in this case the VMCache paper) describe their core algorithms in sufficient detail so we can actually use them as reference. This is very similar to Listing 2 of the paper.

Reply:
Yep! me too.